### ISSUE TITLE
Incorrect unused_imports warning when type is used in generics

### ISSUE THREAD

**Issue by KallDrexx** [rust-lang/rust#23492]
Compiling the following code:

```
use std::io::BufReader;
use std::io::BufRead;
use std::io::Read;
use types;
use fingerprinter;

pub fn read_input<T: Read>(mut reader: BufReader<T>) -> Vec<types::FingerprintedValue> {
    let mut fingerprints = Vec::new();
    let mut words = Vec::new();
    let mut line = String::new();

    let bytes_read = reader.read_line(&mut line).unwrap();
    if bytes_read == 0 {
        panic!("No data in input.txt");
    }

    let mut current_word = String::new();
    for character in line.chars() {
        // Split character on white space
        if character.is_whitespace() {
            if current_word.len() > 0 {
                words.push(current_word);
                current_word = String::new();
                continue;
            }
        }

        current_word.push(character);
    }

    if words.len() > 0 {
        fingerprints.push(types::FingerprintedValue {
            fingerprinted: fingerprinter::get_fingerprint(&words[0][..]),
            original: words[0].clone()
        });
    }

    return fingerprints;
}
```

Gives the following warning:

```
src\input_reader.rs:44:6: 44:19 warning: unused import, #[warn(unused_imports)] on by default
src\input_reader.rs:44  use std::io::Read;
                            ^~~~~~~~~~~~~
```

This warning does not seem correct because `Read` is used in the generic type definition, and if you try to remove the import an errorr occurs that the trait isn't found.

**Comment by alexcrichton** [comment]
Do you have a more standalone example that we could test on (or perhaps the original source)? Unfortunately compiling this example as given yields errors because there are no `types` or `fingerprinter` modules.

**Comment by KallDrexx** [comment]
Woops sorry about that.  I'll try and whip up a stand alone example, our link to the commit in my public repo that reproduces this in a bit.

-----Original Message-----

**Comment by KallDrexx** [comment]
Sorry, this was me being dumb and not reading the warning message correctly.  The warning is actually talking about an import in the test module in the same file, and I failed to notice the line number :(.

**Comment by alexcrichton** [comment]
Aha, no worries! Glad it worked out :)
